dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/JdbcIconStore.java

```diff
@@ -0,0 +1,279 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.commons.util.SqlHelper;
+import org.hisp.dhis.fileresource.FileResourceStore;
+import org.hisp.dhis.user.UserService;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+/**
+ * @author Zubair Asghar
+ */
+@Slf4j
+@Repository("org.hisp.dhis.icon.IconStore")
+@RequiredArgsConstructor
+public class JdbcIconStore implements IconStore {
+
+  private static final ObjectMapper keywordsMapper = new ObjectMapper();
+
+  private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+  private final UserService userService;
+
+  private final FileResourceStore fileResourceStore;
+
+  @Override
+  public long count(IconOperationParams params) {
+
+    String sql = """
+     select count(*) from icon c
+                      """;
+
+    MapSqlParameterSource parameterSource = new MapSqlParameterSource();
+
+    sql = buildIconQuery(params, sql, parameterSource);
+
+    return Optional.ofNullable(
+            namedParameterJdbcTemplate.queryForObject(sql, parameterSource, Long.class))
+        .orElse(0L);
+  }
+
+  @Override
+  public Icon getIconByKey(String key) {
+    final String sql =
+        """
+                select c.iconkey as iconkey, c.description as icondescription, c.keywords as keywords, c.created as created, c.lastupdated as lastupdated,
+                c.fileresourceid as fileresourceid, c.createdby as createdby, c.custom as custom from icon c
+                where iconkey = :key
+                """;
+
+    MapSqlParameterSource params = new MapSqlParameterSource();
+    params.addValue("key", key);
+
+    List<Icon> icons =
+        namedParameterJdbcTemplate.query(
+            sql,
+            params,
+            (rs, rowNum) -> {
+              Icon icon = new Icon();
+              icon.setKey(rs.getString("iconkey"));
+              icon.setDescription(rs.getString("icondescription"));
+              icon.setCustom(rs.getBoolean("custom"));
+              icon.setKeywords(convertKeywordsJsonIntoSet(rs.getString("keywords")));
+              icon.setCreated(rs.getDate("created"));
+              icon.setLastUpdated(rs.getDate("lastupdated"));
+              icon.setCreatedBy(userService.getUser(rs.getLong("createdby")));
+              icon.setFileResource(fileResourceStore.get(rs.getLong("fileresourceid")));
+              return icon;
+            });
+
+    return icons.isEmpty() ? null : icons.get(0);
+  }
+
+  @Override
+  public void save(Icon icon) throws SQLException {
+
+    String sql =
+        "INSERT INTO icon (iconkey,description,keywords,fileresourceid,createdby,created,lastupdated,custom) VALUES (:key,:description,cast(:keywords as jsonb),:fileresourceid,:createdby,now(),now(),:custom) ";
```

Comment:
Why are we using `now()` here? are we not using `setAutoFields()` which already sets these fields? Same in the update

Reply:
because it's convenient. Does it pose any problem?